csrc/keyutils.cpp

```diff
@@ -38,25 +38,48 @@ EVP_PKEY* der2EvpPrivateKey(const unsigned char* der, const int derLen, const bo
   }
 
   if (EVP_PKEY_base_id(result) == EVP_PKEY_RSA) {
-      RSA *rsa = EVP_PKEY_get1_RSA(result);
+      const RSA *rsa = EVP_PKEY_get0_RSA(result);
 
       if (rsa) {
           // We need strip zero CRT values which can confuse OpenSSL
-          BN_null_if_zero(rsa->e);
-          BN_null_if_zero(rsa->p);
-          BN_null_if_zero(rsa->q);
-          BN_null_if_zero(rsa->dmp1);
-          BN_null_if_zero(rsa->dmq1);
-          BN_null_if_zero(rsa->iqmp);
-
-          if (rsa->e) {
-              RSA_blinding_on(rsa, NULL);
-          } else {
-              RSA_blinding_off(rsa);
+          const BIGNUM *n;
+          const BIGNUM *e;
+          const BIGNUM *d;
+          const BIGNUM *p;
+          const BIGNUM *q;
+          const BIGNUM *dmp1;
+          const BIGNUM *dmq1;
+          const BIGNUM *iqmp;
+          bool need_rebuild = false;
+
+          RSA_get0_key(rsa, &n, &e, &d);
+          RSA_get0_factors(rsa, &p, &q);
+          RSA_get0_crt_params(rsa, &dmp1, &dmq1, &iqmp);
+          if (e && BN_is_zero(e)) {
+            need_rebuild = true;
+          } else if (p && BN_is_zero(p)) {
+            need_rebuild = true;
+          } else if (q && BN_is_zero(q)) {
+            need_rebuild = true;
+          } else if (dmp1 && BN_is_zero(dmp1)) {
+            need_rebuild = true;
+          } else if (dmq1 && BN_is_zero(dmq1)) {
+            need_rebuild = true;
+          } else if (iqmp && BN_is_zero(iqmp)) {
+            need_rebuild = true;
           }
 
-          // get1_RSA incremented the key's reference count, so we need to free to decrement it again
-          RSA_free(rsa);
+          if (need_rebuild) {
+            // This key likely only has (n, d) set. Very weird, but it happens in java sometimes.
```

Comment:
If this is the case, where does `e` come from?

Reply:
`e` is set to `0` (as that is how this particular non-standard RSA key is encoded by Java). More importantly, it is ignored provided we never use the key as a public key (which we don't). I'm not 100% certain what would happen if we did try to use the key as a public key.  The important thing about this logic is that we need to unset the CRT parameters. OpenSSL doesn't give a way to clear them, just set them. If they are set (even with the nonsensical value of `0`), OpenSSL will attempt to use them when performing private key operations. This fails. So, by creating a new (clean) RSA key with no CRT parameters set and then explicitly setting `d` and `n` (and the ignored value of `e`), we get a key which can be safely used for private key operations.